### PR TITLE
UCT/TCP: Default device selection, performance traits, and a memory pool.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/uct/tcp/tcp_iface.c [hunk 1]
```diff
@@ -41,48 +45,127 @@ static ucs_status_t uct_tcp_iface_get_address(uct_iface_h tl_iface, uct_iface_ad
...
     attr->ep_addr_len      = 0;
-    attr->cap.flags        = UCT_IFACE_FLAG_CONNECT_TO_IFACE;
-    attr->latency          = 50e-6;       /* 50 usec */
-    attr->bandwidth        = 100e6 / 8.0; /* 100 Mbit/s */
-    attr->overhead         = 50e-6;       /* 50 usec */
-    attr->priority         = 0;
+    attr->cap.flags        = UCT_IFACE_FLAG_CONNECT_TO_IFACE |
+                             UCT_IFACE_FLAG_PENDING;
+
+    attr->cap.am.max_short = 0;
```
Comment (miked-mellanox) [inline src/uct/tcp/tcp_iface.c]: why need set to zero some flags if "memset" does it in line 66?

**Comment by MattBBaker** [comment]
ORNL test failure was pretty comprehensive.
```
[ lots more failed tests truncated. ]
[  FAILED  ] tcp/uct_amo_add_test.add32/0, where GetParam() = tcp/eth0
[  FAILED  ] tcp/uct_amo_add_test.add64/1, where GetParam() = tcp/ib0
[  FAILED  ] tcp/uct_amo_add_test.add64/0, where GetParam() = tcp/eth0
[  FAILED  ] tcp/uct_amo_add_test.add32/1, where GetParam() = tcp/ib0

450 FAILED TESTS
[1483641822.077080] [pod9:12692:0]          async.c:555  UCX  INFO  async handler table is not empty during exit (contains 208 elems)
```

I'm pulling out a few of the specific errors, most are just variations of these.

```
[ RUN      ] tcp/test_ucp_tag_match.send_recv_unexp/1
[1483641280.973606] [pod9:12692:0]    ucp_context.c:357  UCX  INFO  Device all is not available
[1483641280.973700] [pod9:12692:0]        tcp_net.c:58   UCX  ERROR ioctl(req=35142, ifr_name=eth0) failed: Operation not permitted
[1483641280.973721] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4012330 [id=168] ???() ???() not released
[1483641280.973724] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4014220 [id=171] ???() ???() not released
[1483641280.973727] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4011960 [id=167] ???() ???() not released
[1483641280.973729] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400fd70 [id=164] ???() ???() not released
[1483641280.973731] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4014cf0 [id=172] ???() ???() not released
[1483641280.973734] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4013790 [id=170] ???() ???() not released
[1483641280.973736] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400c500 [id=157] ???() ???() not released
[1483641280.973738] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400cc50 [id=158] ???() ???() not released
[1483641280.973741] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4010680 [id=165] ???() ???() not released
[1483641280.973743] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400dbb0 [id=160] ???() ???() not released
[1483641280.973745] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400ec10 [id=162] ???() ???() not released
[1483641280.973748] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4012d40 [id=169] ???() ???() not released
[1483641280.973750] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4016f70 [id=175] ???() ???() not released
[1483641280.973752] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400d3e0 [id=159] ???() ???() not released
[1483641280.973754] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4015800 [id=173] ???() ???() not released
[1483641280.973757] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400a0b0 [id=153] ???() ???() not released
[1483641280.973759] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x3c2dfb0 [id=154] ???() ???() not released
[1483641280.973761] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400f4a0 [id=163] ???() ???() not released
[1483641280.973764] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400be00 [id=156] ???() ???() not released
[1483641280.973766] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4016320 [id=174] ???() ???() not released
[1483641280.973768] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400e3c0 [id=161] ???() ???() not released
[1483641280.973771] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x4010fd0 [id=166] ???() ???() not released
[1483641280.973773] [pod9:12692:0]          async.c:310  UCX  WARN  async 0x3ffecc0 handler 0x400b490 [id=155] ???() ???() not released
[1483641280.973775] [pod9:12692:0]          async.c:311  UCX  WARN  releasing async context with 1 handlers
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/ucp/ucp_test.cc:337: Failure
Error: Input/output error
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/common/test.cc:133: Failure
Failed
Got 24 warnings during the test
[  FAILED  ] tcp/test_ucp_tag_match.send_recv_unexp/1, where GetParam() = \tcp (1 ms)
```

```
[ RUN      ] tcp/uct_p2p_rma_test.put_zcopy/2
[1483641767.219465] [pod9:12692:0]        tcp_net.c:58   UCX  ERROR ioctl(req=35142, ifr_name=ib0) failed: Operation not permitted
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:243: Failure
Error: Input/output error
[  FAILED  ] tcp/uct_p2p_rma_test.put_zcopy/2, where GetParam() = tcp/ib0 (0 ms)
[ RUN      ] tcp/uct_p2p_rma_test.get_bcopy/0
[1483641767.219616] [pod9:12692:0]        tcp_net.c:58   UCX  ERROR ioctl(req=35142, ifr_name=eth0) failed: Operation not permitted
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:243: Failure
Error: Input/output error
[  FAILED  ] tcp/uct_p2p_rma_test.get_bcopy/0, where GetParam() = tcp/eth0 (0 ms)
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/637/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3083/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3085/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/639/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2591/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2589/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/641/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/642/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2598/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3088/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/644/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3090/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/646/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2602/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3092/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/647/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2603/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp pls review

**Comment by shamisp** [comment]
👍

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3111/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3132/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3152/ for details.

**Pull request merged**